### ISSUE TITLE
Make browser globals optional

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#1417]
Currently, axe-core assumes it runs in the browser and thus it assumes it has access to all default browser. Things like `window` and `document` are required for axe-core to run.

This creates a challenge when using axe-core with things like JSDOM. To use axe-core, someone has to set up a number of global variables before they can **include** axe-core. 

```js
const doc = jsdom.jsdom('<!doctype html><html><body></body></html>');
global.document = doc;
global.window = doc.defaultView;

const axe = require('axe-core');
```

This implicit reliance on globals is problematic, and can lead to issues like #1404. A much cleaner solution would be if axe-core would infer its document / window objects from whatever context it is passed:

```js
const results = await axe.run(elm)
```

You can infer an element's document and what the corresponding window is. The only time we would need access to a global object is if `axe.run()` wasn't passed any argument, and ideally, we'd do it only once.

This is a significant piece of work. I think we would have to pass `document` and `window` objects around as arguments to check and matches functions. Alternatively we could also add those two objects to the `context` object. That would be a breaking change, but IMO it is a cleaner solution.

---

An interim solution would be for us to tighten ESLint and only allow `document` and `window` objects as globals. That would prevent a problem like we had in #1404.

**Comment by dylanb** [comment]
Which of the proposals would assist in the work we are doing towards abstracting the DOM dependencies?

**Comment by mohanraj-r** [comment]
Could better context detection also help with [Executing axe in headless chrome yields different results · Issue #2088 · dequelabs/axe-core](https://github.com/dequelabs/axe-core/issues/2088)?

**Comment by padmavemulapati** [comment]
devTask
